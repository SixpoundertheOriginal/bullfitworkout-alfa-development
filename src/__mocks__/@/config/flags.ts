import { vi } from 'vitest';

export const useFeatureFlag = vi.fn((flag: string) => {
  if (flag === 'KPI_ANALYTICS_ENABLED') return true;
  if (flag === 'BW_LOADS_ENABLED') return true;
  return false;
});

export const isFeatureEnabled = useFeatureFlag;

export const FEATURE_FLAGS = {
  KPI_ANALYTICS_ENABLED: true,
  BW_LOADS_ENABLED: true,
};

export default {
  useFeatureFlag,
  isFeatureEnabled,
  FEATURE_FLAGS,
};
