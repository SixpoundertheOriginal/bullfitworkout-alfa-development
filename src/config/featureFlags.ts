// Feature flags for gradual rollout
export const FEATURE_FLAGS = {
  // Analytics derived KPIs (density, rest time analysis, set efficiency)
  ANALYTICS_DERIVED_KPIS_ENABLED: process.env.NODE_ENV === 'development' ||
    process.env.VITE_ANALYTICS_DERIVED_KPIS === 'true',

  // Other existing flags can be added here
} as const;
console.debug('[config/featureFlags] ANALYTICS_DERIVED_KPIS_ENABLED=', FEATURE_FLAGS.ANALYTICS_DERIVED_KPIS_ENABLED);
