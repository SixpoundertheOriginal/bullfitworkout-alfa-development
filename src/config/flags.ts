/**
 * Feature Flag Configuration
 * 
 * Centralized feature flag management for safe rollouts
 */

interface FeatureFlags {
  BW_LOADS_ENABLED: boolean;
  ADVANCED_METRICS: boolean;
  AI_RECOMMENDATIONS: boolean;
}

// Environment-based defaults
const isDevelopment = import.meta.env.DEV;
const isStaging = import.meta.env.VITE_APP_ENV === 'staging';
const isProduction = import.meta.env.VITE_APP_ENV === 'production';

const DEFAULT_FLAGS: FeatureFlags = {
  // Bodyweight loads: enabled in dev/staging, controlled rollout in production
  BW_LOADS_ENABLED: isDevelopment || isStaging || false,
  
  // Advanced metrics features
  ADVANCED_METRICS: isDevelopment || isStaging,
  
  // AI-powered recommendations
  AI_RECOMMENDATIONS: true,
};

// Override flags from environment variables if present
const ENV_OVERRIDES: Partial<FeatureFlags> = {
  BW_LOADS_ENABLED: import.meta.env.VITE_BW_LOADS_ENABLED === 'true',
};

// Merge defaults with environment overrides
export const FEATURE_FLAGS: FeatureFlags = {
  ...DEFAULT_FLAGS,
  ...Object.fromEntries(
    Object.entries(ENV_OVERRIDES).filter(([_, value]) => value !== undefined)
  ) as Partial<FeatureFlags>,
};

export type FeatureFlagKey = keyof FeatureFlags;

/**
 * Hook for accessing feature flags
 */
export function useFeatureFlag(flag: FeatureFlagKey): boolean {
  return FEATURE_FLAGS[flag];
}

/**
 * Utility for checking flags in non-React contexts
 */
export function isFeatureEnabled(flag: FeatureFlagKey): boolean {
  return FEATURE_FLAGS[flag];
}

// Telemetry functions (optional)
export function emitTelemetry(event: string, data: Record<string, any>) {
  if (isDevelopment) {
    console.log(`[Telemetry] ${event}:`, data);
  }
  
  // In production, send to analytics service
  // analytics.track(event, data);
}

export function emitBwLoadApplied(data: {
  exerciseId: string;
  userBw: number;
  loadPerRep: number;
}) {
  emitTelemetry('bw_load_applied', data);
}

export function emitIsometricWorkLogged(data: {
  exerciseId: string;
  loadKg: number;
  durationSec: number;
  workKgSec: number;
}) {
  emitTelemetry('isometric_work_logged', data);
}