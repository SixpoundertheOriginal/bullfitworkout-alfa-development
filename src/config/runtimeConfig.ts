import React, { createContext, useContext, useMemo } from 'react';

export type Flags = {
  derivedKpis: boolean;
};

const FLAG_PATHS: Record<keyof Flags, string> = {
  derivedKpis: 'analytics.derivedKpis.enabled',
};

const isBrowser = typeof window !== 'undefined';

function readLocal(): Record<string, unknown> {
  if (!isBrowser) return {};
  try {
    const raw = window.localStorage.getItem('bf.flags');
    return raw ? JSON.parse(raw) : {};
  } catch {
    return {};
  }
}

function readWindow(): Record<string, unknown> {
  if (!isBrowser) return {};
  return (window as any).__FLAGS__ || {};
}

function readEnv(key: string): unknown {
  const envKey = `VITE_${key.replace(/\./g, '_').toUpperCase()}`;
  // import.meta may not exist in some test contexts
  const env = (import.meta as any)?.env;
  return env ? env[envKey] : undefined;
}

export function getFlag(name: keyof Flags, defaults: Flags): boolean {
  const path = FLAG_PATHS[name];
  const winVal = readWindow()[path];
  if (typeof winVal === 'boolean') return winVal;
  const localVal = readLocal()[path];
  if (typeof localVal === 'boolean') return localVal;
  const envVal = readEnv(path);
  if (typeof envVal === 'string') {
    return envVal === 'true';
  }
  return defaults[name];
}

const ConfigCtx = createContext<{ flags: Flags }>({ flags: { derivedKpis: false } });

export interface ConfigProviderProps {
  children: React.ReactNode;
  initialFlags?: Partial<Flags>;
}

export const ConfigProvider: React.FC<ConfigProviderProps> = ({ children, initialFlags }) => {
  const flags = useMemo<Flags>(() => {
    const defaults: Flags = { derivedKpis: import.meta.env.MODE !== 'production' };
    return {
      derivedKpis:
        typeof initialFlags?.derivedKpis === 'boolean'
          ? initialFlags.derivedKpis
          : getFlag('derivedKpis', defaults),
    };
  }, [initialFlags?.derivedKpis]);

  if (isBrowser) {
    (window as any).__FLAGS__ = {
      ...(window as any).__FLAGS__,
      [FLAG_PATHS.derivedKpis]: flags.derivedKpis,
    };
  }

  return React.createElement(ConfigCtx.Provider, { value: { flags } }, children);
};

export function useConfig() {
  return useContext(ConfigCtx);
}

