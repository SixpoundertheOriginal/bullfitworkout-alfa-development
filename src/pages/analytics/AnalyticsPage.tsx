import React from 'react';
import type { PerWorkoutMetrics, TimeSeriesPoint } from '@/services/metrics-v2/dto';
import { fmtKgPerMin, fmtSeconds, fmtRatio } from './formatters';
import { useConfig } from '@/config/runtimeConfig';
import { buildMetricOptions } from './metricOptions';

export type AnalyticsServiceData = {
  perWorkout?: PerWorkoutMetrics[];
  series?: Record<string, TimeSeriesPoint[]>;
  metricKeys?: string[];
};

export type AnalyticsPageProps = { data?: AnalyticsServiceData };

export const AnalyticsPage: React.FC<AnalyticsPageProps> = ({ data }) => {
  const perWorkout = data?.perWorkout ?? [];
  const seriesData = data?.series ?? {};
  const metricKeys = data?.metricKeys ?? [];
  const { flags } = useConfig();

  React.useEffect(() => {
    console.debug('[AnalyticsPage] render, derivedKpis=', flags.derivedKpis);
  }, [flags.derivedKpis]);

  const initOpts = React.useMemo(
    () => buildMetricOptions(metricKeys, flags.derivedKpis),
    [metricKeys, flags.derivedKpis]
  );
  const [options, setOptions] = React.useState(initOpts);
  const [metric, setMetric] = React.useState<string>(initOpts[0]?.key || '');

  React.useEffect(() => {
    const next = buildMetricOptions(metricKeys, flags.derivedKpis);
    setOptions(next);
    if (!next.some(o => o.key === metric) && next[0]) {
      setMetric(next[0].key);
    }
    if (import.meta.env.DEV && typeof window !== 'undefined') {
      (window as any).__BF_DEBUG__ = {
        ...(window as any).__BF_DEBUG__,
        metricOptions: next,
      };
    }
  }, [metricKeys, flags.derivedKpis]);

  const series = React.useMemo(() => {
    return seriesData[metric] ?? [];
  }, [metric, seriesData]);

  // KPI calculations (last 7 days vs previous 7 days) using server-provided KPIs
  const kpis = React.useMemo(() => {
    const now = new Date('2023-01-15'); // deterministic for tests; replace in real app
    const dayMs = 86400000;
    const startCurr = new Date(now.getTime() - 7 * dayMs);
    const startPrev = new Date(now.getTime() - 14 * dayMs);

    const currWorkouts = perWorkout.filter(w => {
      const d = new Date(w.startedAt);
      return d >= startCurr && d < now;
    });
    const prevWorkouts = perWorkout.filter(w => {
      const d = new Date(w.startedAt);
      return d >= startPrev && d < startCurr;
    });

    const avgKpi = (wps: PerWorkoutMetrics[], key: keyof NonNullable<PerWorkoutMetrics['kpis']>) => {
      const arr = wps
        .map(w => w.kpis?.[key] as number | null | undefined)
        .filter((n): n is number => n !== null && n !== undefined);
      return arr.length > 0 ? arr.reduce((s, n) => s + n, 0) / arr.length : key === 'setEfficiency' ? null : 0;
    };

    const curr = {
      density: avgKpi(currWorkouts, 'density'),
      avgRest: avgKpi(currWorkouts, 'avgRest'),
      efficiency: avgKpi(currWorkouts, 'setEfficiency'),
    };
    const prev = {
      density: avgKpi(prevWorkouts, 'density'),
      avgRest: avgKpi(prevWorkouts, 'avgRest'),
      efficiency: avgKpi(prevWorkouts, 'setEfficiency'),
    };
    return { curr, prev };
  }, [perWorkout]);

  const renderDelta = (value: number, formatter: (n: number) => string) => {
    const sign = value >= 0 ? '+' : '-';
    return sign + formatter(Math.abs(value));
  };

  return (
    <div>
      {flags.derivedKpis && (
        <div className="flex gap-4 mb-4">
          <div data-testid="kpi-density">
            <div>Density</div>
            <div>{fmtKgPerMin(kpis.curr.density)}</div>
            <div>{renderDelta(kpis.curr.density - kpis.prev.density, fmtKgPerMin)}</div>
          </div>
          <div data-testid="kpi-rest">
            <div>Avg Rest</div>
            <div>{fmtSeconds(kpis.curr.avgRest)}</div>
            <div>{renderDelta(kpis.curr.avgRest - kpis.prev.avgRest, fmtSeconds)}</div>
          </div>
          <div data-testid="kpi-efficiency">
            <div>Set Efficiency</div>
            <div>{fmtRatio(kpis.curr.efficiency)}</div>
            <div>{fmtRatio(
              kpis.curr.efficiency !== null && kpis.prev.efficiency !== null
                ? kpis.curr.efficiency - kpis.prev.efficiency
                : null
            )}</div>
          </div>
        </div>
      )}

        <select
          value={metric}
          onChange={e => setMetric(e.target.value)}
          data-testid="metric-select"
        >
        {options.map(o => (
          <option key={o.key} value={o.key}>
            {o.label}
          </option>
        ))}
      </select>

      <pre data-testid="series">{JSON.stringify(series, null, 2)}</pre>
    </div>
  );
};

export default AnalyticsPage;
