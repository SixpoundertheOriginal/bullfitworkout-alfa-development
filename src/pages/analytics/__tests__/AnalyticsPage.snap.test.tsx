import React from 'react';
import { render } from '@testing-library/react';
import { describe, it, expect } from 'vitest';
import type { PerWorkoutMetrics, TimeSeriesPoint } from '@/services/metrics-v2/dto';
import { AnalyticsPage } from '../AnalyticsPage';
import { ConfigProvider } from '@/config/runtimeConfig';

const makeWorkouts = (): PerWorkoutMetrics[] => {
  const workouts: PerWorkoutMetrics[] = [];
  for (let i = 1; i <= 7; i++) {
    workouts.push({
      workoutId: `p${i}`,
      startedAt: `2023-01-${String(i).padStart(2, '0')}T10:00:00Z`,
      totalVolumeKg: 100,
      totalSets: 10,
      totalReps: 0,
      durationMin: 10,
      restMin: 10,
      kpis: { density: 10, avgRest: 60, setEfficiency: 1.0 },
    });
  }
  for (let i = 8; i <= 14; i++) {
    workouts.push({
      workoutId: `c${i}`,
      startedAt: `2023-01-${String(i).padStart(2, '0')}T10:00:00Z`,
      totalVolumeKg: 200,
      totalSets: 10,
      totalReps: 0,
      durationMin: 10,
      restMin: 5,
      kpis: { density: 20, avgRest: 30, setEfficiency: 0.8 },
    });
  }
  return workouts;
};

describe('AnalyticsPage KPI cards', () => {
  it('renders KPI cards when enabled', () => {
    const workouts = makeWorkouts();
    const data = {
      perWorkout: workouts,
      series: { volume: [], sets: [], workouts: [], duration: [], reps: [], density: [], avgRest: [], setEfficiency: [] } as Record<string, TimeSeriesPoint[]>,
      metricKeys: ['volume','sets','workouts','duration','reps','density','avgRest','setEfficiency'],
    };
    const { container, getByTestId } = render(
      <ConfigProvider initialFlags={{ derivedKpis: true }}>
        <AnalyticsPage data={data} />
      </ConfigProvider>
    );
    expect(getByTestId('kpi-density')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('hides KPI cards when disabled', () => {
    const workouts = makeWorkouts();
    const data = {
      perWorkout: workouts,
      series: { volume: [], sets: [], workouts: [], duration: [], reps: [], density: [], avgRest: [], setEfficiency: [] } as Record<string, TimeSeriesPoint[]>,
      metricKeys: ['volume','sets','workouts','duration','reps','density','avgRest','setEfficiency'],
    };
    const { queryByTestId } = render(
      <ConfigProvider initialFlags={{ derivedKpis: false }}>
        <AnalyticsPage data={data} />
      </ConfigProvider>
    );
    expect(queryByTestId('kpi-density')).toBeNull();
  });
});
