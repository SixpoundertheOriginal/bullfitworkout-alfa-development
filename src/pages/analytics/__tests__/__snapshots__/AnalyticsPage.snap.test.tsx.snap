// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AnalyticsPage KPI cards > renders KPI cards when enabled 1`] = `
<div>
  <div>
    <div
      class="flex gap-4 mb-4"
    >
      <div
        data-testid="kpi-density"
      >
        <div>
          Density
        </div>
        <div>
          20.00 kg/min
        </div>
        <div>
          +10.00 kg/min
        </div>
      </div>
      <div
        data-testid="kpi-rest"
      >
        <div>
          Avg Rest
        </div>
        <div>
          0m 30s
        </div>
        <div>
          -0m 30s
        </div>
      </div>
      <div
        data-testid="kpi-efficiency"
      >
        <div>
          Set Efficiency
        </div>
        <div>
          0.80×
        </div>
        <div>
          -0.20×
        </div>
      </div>
    </div>
    <select
      data-testid="metric-select"
    >
      <option
        value="volume"
      >
        Total Volume
      </option>
      <option
        value="sets"
      >
        Total Sets
      </option>
      <option
        value="workouts"
      >
        Workouts
      </option>
      <option
        value="duration"
      >
        Total Duration
      </option>
      <option
        value="reps"
      >
        Total Reps
      </option>
      <option
        value="density"
      >
        Workout Density (kg/min)
      </option>
      <option
        value="avgRest"
      >
        Avg Rest / Session (sec)
      </option>
      <option
        value="setEfficiency"
      >
        Set Efficiency (×)
      </option>
    </select>
    <pre
      data-testid="series"
    >
      []
    </pre>
  </div>
</div>
`;
